### PR TITLE
More robust zip file handling

### PR DESCRIPTION
**Pull request by wagnerand**
- Stop calling SafeZip(upload.path).is_valid() in handle_upload_validation_result(). Keep the try/except around the check_for_api_keys_in_file() call though. Modify the exept to catch ValidationError, BadZipFile, IOError.
- In check_for_api_keys_in_file() when you get the zip_file instance, call is_valid() on it before doing anything.
- Change for filepath in zipfile.filelist to for zipinfo in zipfile.info_list (we build that property inside is_valid())
- Change filepath to zipinfo everywhere (it never was just a filepath)
- In the loop, skip everything if zipinfo.file_size is == 0
- In the loop, decode the file contents with .decode('unicode', errors='ignore')
- Add test to invalid zip (that would trigger a ValidationError we didn't catch before) and for binary content inside zip (that would trigger the UnicodeDecodeError we didn't catch before)

### CODE REVIEW

**Comment by wagnerand** [comment]
@diox Questions I have:

* Which `ValidationError` can be thrown?. `django.core.validators` or `django.core.exceptions`?
* What would a zip file need to look like to throw `ValidationError`?

**Comment by EnTeQuAk** [comment]
> Which ValidationError can be thrown? django.core.validators or django.core.exceptions?

They're the same, `django.core.validators` just imports `django.core.exceptions`

> What would a zip file need to look like to throw ValidationError?

Can you expand that question a bit, what do you mean? Raising a `ValidationError` is totally valid to do and should not result in a server error.

**Comment by wagnerand** [comment]
> Can you expand that question a bit, what do you mean? Raising a ValidationError is totally valid to do and should not result in a server error.

Mat gave me the todo list in the initial comment, and asked me to test an invalid zip that would trigger a `ValidationError`. I am not sure what such a zip would look like.

**Comment by EnTeQuAk** [comment]
> Mat gave me the todo list in the initial comment, and asked me to test an invalid zip that would trigger a ValidationError. I am not sure what such a zip would look like.

Anything that disguises as a zip would work (rename a text file to .zip and try, or rename a tar.gz to a .zip file) imho.

File: src/olympia/devhub/tasks.py [hunk 1]
```diff
@@ -451,67 +445,50 @@ def annotate_webext_incompatibilities(results, file_, addon, version_string,
...
-
-                    # Revoke and regenerate after 2 minutes to allow the
-                    # developer to fetch the validation results
-                    log.info('X-awagner check_for_api_keys_in_file : Calling '
-                             'revoke routine')
-                    revoke_and_regenerate_api_key.apply_async(
-                        kwargs={'key_id': key.id}, countdown=120)
+        zipfile.is_valid()
+        for zipinfo in zipfile.info_list:
+            if zipinfo.file_size > 0:
```
Comment (eviljeff) [inline src/olympia/devhub/tasks.py]: nit/suggestion - could be `>=` whatever the length is for private keys. 64 bytes?

File: src/olympia/devhub/tasks.py [hunk 2]
```diff
@@ -451,67 +445,50 @@ def annotate_webext_incompatibilities(results, file_, addon, version_string,
...
         users = [upload.user] if upload.user else []
-    log.info('X-awagner check_for_api_keys_in_file : checking keys for users '
-             '%s' % users)
 
     keys = []
     for user in users:
         try:
-            log.info('X-awagner check_for_api_keys_in_file : Getting key for '
-                     'user %s' % user)
             key = APIKey.get_jwt_key(user_id=user.id)
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/devhub/tasks.py]: Yes, I'll file a follow-up issue.

**Pull request merged**